functions/geometry_vi

```diff
@@ -0,0 +1,77 @@
+# geometry_vi - A geometry plugin to display the ZLE vi-mode state.
+
+geometry_vi() {
+    # Required for RPROMPT to render properly after submitting a command.
+    geometry::vi-get-mode
+}
+
+# Re-render the prompt.
+function geometry::vi-update-prompt() {
+    imode=$(geometry::vi-insert-mode)
+    nmode=$(geometry::vi-normal-mode)
+    if [[ $KEYMAP == vicmd ]]; then
+        PROMPT=${PROMPT//$imode/$nmode}
+        RPROMPT=${RPROMPT//$imode/$nmode}
+    else
+        PROMPT=${PROMPT//$nmode/$imode}
+        RPROMPT=${RPROMPT//$nmode/$imode}
+    fi
+}
+
+# Render normal mode.
+function geometry::vi-normal-mode() {
+    ansi ${GEOMETRY_VI_NORMAL_COLOR:-white} ${GEOMETRY_VI_NORMAL_MODE:-"<NORMAL>"}
+}
+
+# Render insert mode.
+function geometry::vi-insert-mode() {
+    ansi ${GEOMETRY_VI_INSERT_COLOR:-yellow} ${GEOMETRY_VI_INSERT_MODE:-"<INSERT>"}
+}
+
+# Return the currently active mode.
+function geometry::vi-get-mode() {
+    imode=$(geometry::vi-insert-mode)
+    nmode=$(geometry::vi-normal-mode)
+    [[ $KEYMAP == vicmd ]] && echo $nmode || echo $imode
+}
+
+# ZLE widget to trigger prompt updates.
+function geometry::vi-draw-mode {
+    keymap=$(geometry::vi-get-mode)
+    geometry::vi-update-prompt
+    zle reset-prompt
+}
+
+
+# Initialization: Bind the widgets required to perform proper prompt updates.
```

Comment:
Thanks to the changes in #273, this will only run once if `geometry_vi` is included in any of the rendering prompts.